### PR TITLE
Add SAG implementation

### PR DESCRIPTION
**Pull request by superchild**
#### What I did
* Add static anycast gateway related commands support. 
  Refer to [SAG HLD#837](https://github.com/Azure/SONiC/pull/837)

#### How I did it
* Add new commands
    * config static-anycast-gateway mac_address add/del
    * config vlan static-anycast-gateway add/del
* Add unit test cases

#### How to verify it
* Executing unit tests

Signed-off-by: Jimi Chen jimi_chen@edge-core.com

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -5852,6 +5852,47 @@ def disable_link_local(ctx):
...
+    pass
+
+@mac_address.command('add')
+@click.argument('mac_address', metavar='<mac_address>', required=True, type=str)
+@clicommon.pass_db
+def add_mac(db, mac_address):
+    """Add static-anycast-gateway mac address command"""
+    log.log_info(f"'static-anycast-gateway mac_address add {mac_address}' executing...")
+
+    if not db.cfgdb.get_entry('SAG', 'GLOBAL'):
```
Comment (cbpaviz) [inline config/main.py:5877]: nit: Is it possible to add a check to validate the MAC address (to be only unicast)

File: config/vlan.py [hunk 2]
```diff
@@ -185,3 +185,45 @@ def del_vlan_member(db, vid, port):
...
+    pass
+
+@static_anycast_gateway.command('add')
+@click.argument('vid', metavar='<vid>', required=True, type=int)
+@clicommon.pass_db
+def add_vlan_sag(db, vid):
+    """Enable static-anycast-gatweay on VLAN interface"""
+    ctx = click.get_current_context()
+
+    log.log_info(f"'vlan static-anycast-gateway add {vid}' executing...")
```
Comment (cbpaviz) [inline config/vlan.py:202]: Can this command be changed to "config vlan <vid> static-anycast-gateway enable" and ""config vlan <vid> static-anycast-gateway disable"

File: show/main.py [hunk 3]
```diff
@@ -1689,6 +1689,22 @@ def ztp(status, verbose):
...
     run_command(cmd, display_cmd=verbose)
 
+#
+# 'static anycast gateway' command ("show static-anycast-gateway")
+#
+@cli.command('static-anycast-gateway')
+@clicommon.pass_db
+def sag(db):
+    """Show static anycast gateway information"""
+    sag_entry = db.cfgdb.get_entry('SAG', 'GLOBAL')
```
Comment (cbpaviz) [inline show/main.py:1699]: Can "show static-anycast-gateway" also show the list of VLANs for which static-anycast-gateway is enabled. This will eliminate the need for 2 commands to be executed to check the configuration

File: config/main.py (hunk 1 above)
Comment (superchild) [reply config/main.py:5885]: Added MAC address format checking and also only allow unicast MAC address.

File: config/vlan.py (hunk 2 above)
Comment (superchild) [reply config/vlan.py]: Changed the corresponding commands from add/del to enable/disable.

File: show/main.py (hunk 3 above)
Comment (superchild) [reply show/main.py:1699]: Added static anycast gateway enabled VLAN interfaces in the display format.

**Comment by superchild** [comment]
@cbpaviz 
Thanks for your comments, codes are modified according that.
Please help to check and review again.

**Review by cbpaviz — DISMISSED** [review]

**Comment by superchild** [comment]
@cbpaviz Could you help to review again, I merged the code from master branch due to some confict.

**Comment by superchild** [comment]
@cbpaviz 
 Could you help to review again?
 Thanks.

**Review by cbpaviz — DISMISSED** [review]

**Comment by superchild** [comment]
@cbpaviz Could you help to review again, I merged the code from master branch due to some conflict.

**Comment by kulpatel** [comment]
@superchild , why SAG config commands are different compare to EC config commands (available in their config guid). is it intentional ?

**Comment by superchild** [comment]
@kulpatel, this PR is discussed and reviewed by community, EC will also follow the same design after this merged. 
The original design is slightly different from community's opinion.

File: config/main.py [hunk 4]
```diff
@@ -6879,6 +6879,55 @@ def disable_link_local(ctx):
...
+        db.cfgdb.set_entry('SAG', 'GLOBAL', {'gateway_mac': mac_address})
+    else:
+        click.get_current_context().fail(f'static-anycast-gateway MAC address {mac_address} is alreday existed. Remove it first')
+
+@mac_address.command('del')
+@click.argument('mac_address', metavar='<mac_address>', required=True, type=str)
+@clicommon.pass_db
+def del_mac(db, mac_address):
```
Comment (Junchao-Mellanox) [inline config/main.py:6920]: Since only one mac address is allowed. mac_address is not a useful argument here.

File: config/vlan.py [hunk 5]
```diff
@@ -95,6 +95,12 @@ def del_vlan(db, vid):
     # We need to restart dhcp_relay service after dhcpv6_relay config change
     dhcp_relay_util.handle_restart_dhcp_relay_service()
 
+    keys = [ (k, v) for k, v in db.cfgdb.get_table('VLAN_MEMBER') if k == 'Vlan{}'.format(vid) ]
```
Comment (Junchao-Mellanox) [inline config/vlan.py:98]: is this change related to this feature?

File: config/vlan.py [hunk 6]
```diff
@@ -232,3 +238,45 @@ def del_vlan_member(db, vid, port):
...
+    """Enable static-anycast-gatweay on VLAN interface"""
+    ctx = click.get_current_context()
+
+    log.log_info(f"'vlan static-anycast-gateway enable {vid}' executing...")
+
+    vlan = f'Vlan{vid}'
+    if not clicommon.is_valid_vlan_interface(db.cfgdb, vlan):
+        ctx.fail(f"Interface {vlan} does not exist")
+
+    db.cfgdb.mod_entry('VLAN_INTERFACE', vlan, {"static_anycast_gateway": "true"})
```
Comment (Junchao-Mellanox) [inline config/vlan.py:261]: do we need check whether static_anycast_gateway has already been configured?

File: config/main.py [hunk 7]
```diff
@@ -6879,6 +6879,55 @@ def disable_link_local(ctx):
...
+    """Del static-anycast-gateway mac address command"""
+    log.log_info(f"'static-anycast-gateway mac_address del {mac_address}' executing...")
+
+    sag_entry = db.cfgdb.get_entry('SAG', 'GLOBAL')
+    if sag_entry:
+        if sag_entry.get('gateway_mac').lower() == mac_address.lower():
+            db.cfgdb.mod_entry('SAG', 'GLOBAL', None)
```
Comment (Junchao-Mellanox) [inline config/main.py:6927]: do we need disable sag for each vlan?

File: config/vlan.py [hunk 8]
```diff
@@ -232,3 +238,45 @@ def del_vlan_member(db, vid, port):
...
+@static_anycast_gateway.command('disable')
+@click.argument('vid', metavar='<vid>', required=True, type=int)
+@clicommon.pass_db
+def disable_vlan_sag(db, vid):
+    """Disable static-anycast-gatweay on VLAN interface"""
+    ctx = click.get_current_context()
+
+    log.log_info(f"'vlan static-anycast-gateway disable {vid}' executing...")
+
+    if not clicommon.is_vlanid_in_range(vid):
```
Comment (Junchao-Mellanox) [inline config/vlan.py:274]: if this check is needed, maybe we should do the same for enable_vlan_sag; otherwise, please remove this check

File: config/vlan.py [hunk 9]
```diff
@@ -95,6 +95,12 @@ def del_vlan(db, vid):
     # We need to restart dhcp_relay service after dhcpv6_relay config change
     dhcp_relay_util.handle_restart_dhcp_relay_service()
 
+    keys = [ (k, v) for k, v in db.cfgdb.get_table('VLAN_MEMBER') if k == 'Vlan{}'.format(vid) ]
+
+    if keys:
+        ctx.fail("VLAN ID {} can not be removed. First remove all members assigned to this VLAN.".format(vid))
```
Comment (Junchao-Mellanox) [inline config/vlan.py:101]: Use f-string

File: config/vlan.py [hunk 10]
```diff
@@ -95,6 +95,12 @@ def del_vlan(db, vid):
...
+
+    db.cfgdb.set_entry('VLAN', 'Vlan{}'.format(vid), None)
```
Comment (Junchao-Mellanox) [inline config/vlan.py:103]: Use f-string

File: show/main.py [hunk 11]
```diff
@@ -2028,6 +2028,29 @@ def ztp(status, verbose):
...
```
Comment (Junchao-Mellanox) [inline show/main.py:2038]: what if SAG is not configured?

File: show/main.py [hunk 12]
```diff
@@ -2028,6 +2028,29 @@ def ztp(status, verbose):
...
+    sag_mac = sag_entry.get('gateway_mac')
+    
+    header = ['MacAddress', 'Interfaces']
+    body = []
+    intf_dict = db.cfgdb.get_table('VLAN_INTERFACE')
+    if intf_dict:
+        for intf in intf_dict:
```
Comment (Junchao-Mellanox) [inline show/main.py:2045]: It would be more brief to use
```python
for key,value in intf_dict.items():
    if value.get('static_anycast_gateway') == 'true':
```

File: show/vlan.py [hunk 13]
```diff
@@ -89,6 +89,18 @@ def get_proxy_arp(ctx, vlan):
 
     return proxy_arp
 
+def get_static_anycast_gateway(ctx, vlan):
+    cfg, _ = ctx
+    _, vlan_ip_data, _ = cfg
+
+    for key in vlan_ip_data:
```
Comment (Junchao-Mellanox) [inline show/vlan.py:96]: why do we need a for loop here? how about this:
```python
if vlan in vlan_ip_data:
```

**Review by Junchao-Mellanox — CHANGES_REQUESTED** [review]

**Comment by superchild** [comment]
@zhangyanzhao 
Sorry for the wrong operation sync, this PR's commit history is gone.
I already reset the commit on my dev branch, could you please reopen this PR?
Or should I resend a new PR for it?

File: config/main.py (hunk 4 above)
Comment (superchild) [reply config/main.py]: This is removed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: config/vlan.py (hunk 5 above)
Comment (superchild) [reply config/vlan.py]: This is reverted in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: config/vlan.py (hunk 6 above)
Comment (superchild) [reply config/vlan.py]: This is removed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: config/main.py (hunk 7 above)
Comment (superchild) [reply config/main.py]: This is handled in sonic-swss.

File: config/vlan.py (hunk 8 above)
Comment (superchild) [reply config/vlan.py]: This is removed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: show/main.py (hunk 11 above)
Comment (superchild) [reply show/main.py]: This is changed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: show/main.py (hunk 12 above)
Comment (superchild) [reply show/main.py]: This is removed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881

File: show/vlan.py (hunk 13 above)
Comment (superchild) [reply show/vlan.py]: This is removed in the updated PR https://github.com/sonic-net/sonic-utilities/pull/2881
